SCons/Subst.py

```diff
@@ -593,8 +600,11 @@ def expand(self, s, lvars, within_list):
             # SCons has the unusual Null class where any __getattr__ call returns it's self, 
             # which does not work the signature module, and the Null class returns an empty
             # string if called on, so we make an exception in this condition for Null class
-            if (isinstance(s, SCons.Util.Null) or
-                set(signature(s).parameters.keys()) == set(['target', 'source', 'env', 'for_signature'])):
+            # Also allow callables where the only non default valued args match the expected defaults
+            # this should also allow functools.partial's to work.
+            if isinstance(s, SCons.Util.Null) or {k for k, v in signature(s).parameters.items() if
```

Comment:
Maybe just use a `filter` here? You're only checking and using the key, after all, and this implies creating a temporary `dict` to do the check. Using a `filter` should perform better than that.

Reply:
We're using both the key and the value.. 
The key to see if the parameter name is in the list of default expected args, and the value to determine if it doesn't have a default a value.

Also items() doesn't create a dictionary, it returns a view object. So no dict should get created. See https://docs.python.org/3/library/stdtypes.html#dict-views and https://docs.python.org/3/library/stdtypes.html#typesmapping (then scroll down to items())

Reply:
Ah, no, we're talking about two different things, and I was talking about something that doesn't exist in your code. I was for some reason seeing a dictionary comprehension rather than a set comprehension for some reason. I still think a `filter` might be a somewhat prettier way to do it (in my obviously biased opinion), but in the end you're doing a whole comparison anyway, so it doesn't matter performance-wise.